custom_ops.py

```diff
@@ -0,0 +1,78 @@
+import torch
+import numpy as np
+# aka torch.library
+from library import OpDef, traceable, device_types
+
+# =====================================================================
+# This is example user library code. It defines
+# my_sin and my_sin_cos operators.
+# =====================================================================
+
+# User provides their custom op schema and implementations
+class MySin(OpDef):
+    schema = "(Tensor x) -> Tensor"
+
+    # the black-box cpu kernel
+    @staticmethod
+    def impl_cpu(x):
+        return torch.from_numpy(np.sin(x.detach().cpu().numpy()))
+
+    # the black-box cuda kernel
+    @staticmethod
+    def impl_cuda(x):
+        return torch.from_numpy(np.sin(x.detach().cpu().numpy())).to(x.device)
+
+    # the abstract impl. Must be "traceable". User must use opcheck to test.
+    @staticmethod
+    def abstract(x):
+        return torch.empty_like(x)
+
+    # autograd: provide us setup_backward() and backward() methods.
+    # these must be "traceable". User must use opcheck to test.
+    @staticmethod
+    def setup_backward(ctx, inputs, output):
+        x, = inputs
+        ctx.save_for_backward(x)
+        return output
+
+    @staticmethod
+    def backward(ctx, grad_output):
+        x, = ctx.saved_tensors 
+        return grad_output * x.cos()
+
+
+# Builds the provided implementations into a "custom op" that behaves like a function.
+# We don't do something like MySin.apply (i.e. the autograd.Function way) because
+# dynamic changes to methods on the MySin object will not be included in the op.
+my_sin_op = MySin.build()
```

Comment:
So, is this the *real* reason you did it this way? Or was it because you need to run some extra code after the class def. An alternate implementation strategy that doesn't require build() is a custom metaclass for OpDef, which performs the initialization work after the class is done being constructed.

It's not clear to me that the justification here for build holds much water. You could always just make sure you getattr(MySin, ...) inside of the registered dispatcher functions so you always are using the up to date function. Or if you close over MySin, this will happen automatically.

Reply:
This is the real reason why I did it this way, yes. I considered the metaclass approach. That seemed a bit too magical. I tried to explain this (the metaclass thing) to some folks once and it blew their minds.

I'm not completely opposed to it though

Reply:
> It's not clear to me that the justification here for build holds much water. You could always just make sure you getattr(MySin, ...) inside of the registered dispatcher functions so you always are using the up to date function. Or if you close over MySin, this will happen automatically.

So the problem is that build() is potentially expensive. For example, if someone updates their MySin object with a `@traceable(impl)`, then we need to figure that out in all the rules we generate from it (e.g. abstract impl, vmap rule). It seemed more explicit to have a build() function (the model is that the user provides implementations and then builds a custom op out of them).